kafka-plugins-0.10/src/main/java/co/cask/hydrator/plugin/source/KafkaStreamingSource.java

```diff
@@ -0,0 +1,285 @@
+@Name("Kafka")
+@Description("Kafka streaming source.")
+public class KafkaStreamingSource extends ReferenceStreamingSource<StructuredRecord> {
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaStreamingSource.class);
+  private final KafkaConfig conf;
+
+  public KafkaStreamingSource(KafkaConfig conf) {
+    super(conf);
+    this.conf = conf;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) throws IllegalArgumentException {
+    super.configurePipeline(pipelineConfigurer);
+    conf.validate();
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(conf.getSchema());
+    if (conf.getMaxRatePerPartition() != null && conf.getMaxRatePerPartition() > 0) {
+      Map<String, String> pipelineProperties = new HashMap<>();
+      pipelineProperties.put("spark.streaming.kafka.maxRatePerPartition", conf.getMaxRatePerPartition().toString());
+      pipelineConfigurer.setPipelineProperties(pipelineProperties);
+    }
+  }
+
+  @Override
+  public JavaDStream<StructuredRecord> getStream(StreamingContext context) throws Exception {
+    context.registerLineage(conf.referenceName);
+
+    Map<String, Object> kafkaParams = new HashMap<>();
+    kafkaParams.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, conf.getBrokers());
+    // Spark saves the offsets in checkpoints, no need for Kafka to save them
+    kafkaParams.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false");
+    kafkaParams.put("key.deserializer", ByteArrayDeserializer.class.getCanonicalName());
+    kafkaParams.put("value.deserializer", ByteArrayDeserializer.class.getCanonicalName());
+    KafkaHelpers.setupKerberosLogin(kafkaParams, conf.getPrincipal(), conf.getKeytabLocation());
+    // Create a unique string for the group.id using the pipeline name and the topic
+    kafkaParams.put("group.id", Joiner.on("-").join(context.getPipelineName().length(), conf.getTopic().length(),
+                                                    context.getPipelineName(), conf.getTopic()));
+    kafkaParams.putAll(conf.getKafkaProperties());
+
+    Properties properties = new Properties();
+    properties.putAll(kafkaParams);
+    try (Consumer<byte[], byte[]> consumer = new KafkaConsumer<>(properties, new ByteArrayDeserializer(),
+                                                                 new ByteArrayDeserializer())) {
+      Map<TopicPartition, Long> offsets = conf.getInitialPartitionOffsets(getPartitions(consumer));
+      // KafkaUtils doesn't understand -1 and -2 as smallest offset and latest offset.
+      // so we have to replace them with the actual smallest and latest
+      List<TopicPartition> earliestOffsetRequest = new ArrayList<>();
+      List<TopicPartition> latestOffsetRequest = new ArrayList<>();
+      for (Map.Entry<TopicPartition, Long> entry : offsets.entrySet()) {
+        TopicPartition topicAndPartition = entry.getKey();
+        Long offset = entry.getValue();
+        if (offset == OffsetRequest.EarliestTime()) {
```

Comment:
I don't understand this logic. The offsets can have -1 and -2 as their value, if that is the case, this will never match right?

Reply:
`OffsetRequest.EarliestTime()` always returns -2 and `OffsetRequest.LatestTime()` always returns -1, so they will match.

Reply:
👍